bip-0346.md

```diff
@@ -0,0 +1,438 @@
+* The element is interpreted as the TxFieldSelector and is popped off the stack.
+* The TxFieldSelector must be valid, fail otherwise.
+* The 32-byte TxHash of the transaction at the current input index, calculated
+  using the given TxFieldSelector is pushed onto the stack.
+
+## TxFieldSelector
+
+The TxFieldSelector has the following semantics. We will give a brief conceptual
+summary, followed by a reference implementation of the CalculateTxHash function.
+
+* There are two special cases for the TxFieldSelector:
+  * the empty value, zero bytes long: it is set equal to `TXFS_SPECIAL_TEMPLATE`,
+    the de-facto default value which means everything except the prevouts and the prevout
+    scriptPubkeys and amounts.
+
+    Special case `TXFS_SPECIAL_TEMPLATE` is 4 bytes long, as follows:
+    * 1: `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX`
+    * 2: `TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_OUTPUTS_ALL`
+    * 3: `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+    * 4: `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+
+  * If the TxFieldSelector has exactly 1 byte, it has its 8 bits
+    assigned as follows, from lowest to highest:
+    * 2/1: Inputs
+      * 00: `TXFS_INOUT_SELECTION_NONE`
+      * 01: `TXFS_INOUT_SELECTION_CURRENT`
+      * 11: `TXFS_INOUT_SELECTION_ALL`
+    * 4/3: Outputs
+      * 00: `TXFS_INOUT_SELECTION_NONE`
+      * 01: `TXFS_INOUT_SELECTION_CURRENT`
+      * 11: `TXFS_INOUT_SELECTION_ALL`
+    * 5: `TXFS_INPUTS_PREVOUTS`
+    * 6: `TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES`
+    * 7: `TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_SPENTSCRIPT | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`
+    * 8: `TXFS_CURRENT_INPUT_IDX`
+
+    Additionally, it includes `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CONTROL | TXFS_CURRENT_INPUT_TAPROOT_ANNEX | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_OUTPUTS_ALL`
+
+    These 1-byte selections allow the TxFieldSelector to emulate current
+    signature hashing modes and those defined in BIP-118:
+
+| BIP-341/118 sighash type     | 1-byte TxFieldSelector |
+| :--------------------------- | :--------------------- |
+| `ALL`                        | `0b11111111`           |
+| `SINGLE`                     | `0b11110111`           |
+| `NONE`                       | `0b11110011`           |
+| `ALL|ANYONECANPAY`           | `0b11111101`           |
+| `SINGLE|ANYONECANPAY`        | `0b11110101`           |
+| `NONE|ANYONECANPAY`          | `0b11110001`           |
+| `ALL|ANYPREVOUT`             | `0b11101101`           |
+| `SINGLE|ANYPREVOUT`          | `0b11100101`           |
+| `NONE|ANYPREVOUT`            | `0b11100001`           |
+| `ALL|ANYPREVOUTANYSCRIPT`    | `0b11001101`           |
+| `SINGLE|ANYPREVOUTANYSCRIPT` | `0b11000101`           |
+| `NONE|ANYPREVOUTANYSCRIPT`   | `0b11000001`           |
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
```

Comment:
It seems to me that we are now in the case where the TxFieldSelector is longer than 1 byte. It would perhaps be helpful to explicitly communicate that:

```suggestion
* If the TxFieldSelector is longer than 1 byte, the first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
```